lib/propolis/src/enlightenment/hyperv/mod.rs

```diff
@@ -168,6 +208,78 @@ impl HyperV {
             }
         }
     }
+
+    fn handle_rdmsr_time_ref_count(&self) -> RdmsrOutcome {
+        if !self.features.reference_tsc {
+            return RdmsrOutcome::GpException;
+        }
+
+        let time_data = vmm::time::export_time_data(self.vmm_hdl())
+            .expect("VMM time data can always be exported");
+
+        // `hrtime` is the current host high-resolution timer value (in
+        // nanoseconds), and `boot_hrtime` is the high-resolution timer value at
+        // which the VM was started, so it suffices to subtract the latter from
+        // the former and divide by 100 to get 100-nanosecond units.
+        //
+        // This module assumes that if its user migrates this VM, the migration
+        // procedure will adjust `boot_hrtime` on the target to account for any
+        // hrtime differences between the source and target VM hosts.
+        let ns_since_boot: u64 = time_data
+            .hrtime
+            .checked_sub(time_data.boot_hrtime)
+            .expect("overflow/underflow while calculating reference uptime")
```

Comment:
thanks for bearing with me kicking the tires on the arithmetic here - if you want to keep the `.expect()`s, how do you feel about including a note that overflow implies the VM was up `2**63 / 1e9 / 3600 / 24 / 356 years` aka >240 years?

Reply:
Annotated in b3d3a74. I also added a USDT probe to this path to verify that I could get a guest to hit it (and I can: Linux will hammer this MSR if you offer the TSC enlightenment but write a 0 to the `sequence` field of the reference TSC page).

Reply:
ah this is lovely, thanks!